File: analysis/import/CSVImporter/src/main/java/de/maibornwolff/codecharta/importer/csv/CSVImporter.java
package de.maibornwolff.codecharta.importer.csv;

import com.google.common.collect.ImmutableList;
import de.maibornwolff.codecharta.serialization.ProjectSerializer;
import picocli.CommandLine;

import java.io.*;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Callable;
import java.util.stream.Collectors;

@CommandLine.Command(name = "csvimport",
        description = "generates cc.json from csv with header",
        footer = "Copyright(c) 2018, MaibornWolff GmbH"
)
public class CSVImporter implements Callable<Void> {

    @CommandLine.Option(names = {"-h", "--help"}, usageHelp = true, description = "displays this help and exits")
    private Boolean help = false;

    @CommandLine.Option(names = {"-d", "--delimeter"}, description = "delimeter in csv file")
    private char csvDelimiter = ',';

    @CommandLine.Option(names = {"-p", "--projectName"}, description = "project name")
    private String projectName = "SCMLogParser";

    @CommandLine.Option(names = "--pathSeparator", description = "path separator (default = '/')")
    private char pathSeparator = '/';

    @CommandLine.Parameters(arity = "1..*", paramLabel = "FILE", description = "csv files")
    private List<String> files = new ArrayList<>();

    public static void main(String... args) {
        CommandLine.call(new CSVImporter(), System.out, args);
    }

    private static List<InputStream> getInputStreamsFromArgs(List<String> files) {
        List<InputStream> fileList = files.stream().map(CSVImporter::createFileInputStream).collect(Collectors.toList());
        return fileList.isEmpty() ? ImmutableList.of(System.in) : fileList;
    }

    private static FileInputStream createFileInputStream(String path) {
        try {
            return new FileInputStream(path);
        } catch (FileNotFoundException e) {
            throw new RuntimeException("File " + path + " not found.");
        }
    }

    @Override
    public Void call() throws IOException {
        CSVProjectAdapter project = new CSVProjectAdapter(projectName, pathSeparator, csvDelimiter);
        getInputStreamsFromArgs(files).forEach(project::addProjectFromCsv);
        ProjectSerializer.serializeProject(project, new OutputStreamWriter(System.out));

        return null;
    }
}

File: analysis/import/CSVImporter/src/main/kotlin/de/maibornwolff/codecharta/importer/csv/CSVImporter.kt
package de.maibornwolff.codecharta.importer.csv

import de.maibornwolff.codecharta.serialization.ProjectSerializer
import de.maibornwolff.codecharta.tools.interactiveparser.InteractiveParser
import de.maibornwolff.codecharta.tools.interactiveparser.ParserDialogInterface
import picocli.CommandLine
import java.io.File
import java.io.IOException
import java.io.InputStream
import java.io.PrintStream
import java.util.concurrent.Callable

@CommandLine.Command(
    name = "csvimport",
    description = ["generates cc.json from csv with header"],
    footer = ["Copyright(c) 2022, MaibornWolff GmbH"]
)
class CSVImporter(
    private val output: PrintStream = System.out
) : Callable<Void>, InteractiveParser {

    @CommandLine.Option(names = ["-h", "--help"], usageHelp = true, description = ["displays this help and exits"])
    private var help = false

    @CommandLine.Option(names = ["-d", "--delimiter"], description = ["delimiter in csv file"])
    private var csvDelimiter = ','

    @CommandLine.Option(names = ["--path-separator"], description = ["path separator (default = '/')"])
    private var pathSeparator = '/'

    @CommandLine.Option(names = ["-nc", "--not-compressed"], description = ["save uncompressed output File"])
    private var compress = true

    @CommandLine.Option(names = ["-o", "--output-file"], description = ["output File"])
    private var outputFile: String? = null

    @CommandLine.Parameters(arity = "1..*", paramLabel = "FILE", description = ["sourcemonitor csv files"])
    private var files: List<File> = mutableListOf()

    @CommandLine.Option(names = ["--path-column-name"], description = ["name of path column"])
    private var pathColumnName: String = "path"

    @Throws(IOException::class)
    override fun call(): Void? {

        val csvProjectBuilder = CSVProjectBuilder(pathSeparator, csvDelimiter, pathColumnName)
        files.map { it.inputStream() }.forEach<InputStream> { csvProjectBuilder.parseCSVStream(it) }
        val project = csvProjectBuilder.build()

        ProjectSerializer.serializeToFileOrStream(project, outputFile, output, compress)

        return null
    }

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            CommandLine.call(CSVImporter(), System.out, *args)
        }
    }

    override fun getDialog(): ParserDialogInterface = ParserDialog
}
